com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/UploadAppStep.java

```diff
@@ -67,16 +69,27 @@ public StepPhase executeAsyncStep(ExecutionWrapper execution) throws FileStorage
             return StepPhase.DONE;
         }
 
-        getStepLogger().debug(Messages.UPLOADING_FILE_0_FOR_APP_1, fileName, app.getName());
-
         String newApplicationDigest = getNewApplicationDigest(execution, appArchiveId, fileName);
-        detectApplicationFileDigestChanges(execution, app.getName(), client, newApplicationDigest);
+        boolean contentChanged = detectApplicationFileDigestChanges(execution, appName, client, newApplicationDigest);
+        if (!contentChanged && isAppStagedCorrectly(client, appName)) {
+            getStepLogger().info(Messages.CONTENT_OF_APPLICATION_0_IS_NOT_CHANGED, appName);
+            return StepPhase.DONE;
+        }
+
+        getStepLogger().debug(Messages.UPLOADING_FILE_0_FOR_APP_1, fileName, appName);
         UploadToken uploadToken = asyncUploadFiles(execution, client, app, appArchiveId, fileName);
-        getStepLogger().debug(Messages.STARTED_ASYNC_UPLOAD_OF_APP_0, app.getName());
+
+        getStepLogger().debug(Messages.STARTED_ASYNC_UPLOAD_OF_APP_0, appName);
         StepsUtil.setUploadToken(uploadToken, execution.getContext());
         return StepPhase.POLL;
     }
 
+    private boolean isAppStagedCorrectly(CloudControllerClient client, String appName) {
+        ApplicationStager appStager = new ApplicationStager(client);
+        CloudApplication cloudApp = client.getApplication(appName);

```

Comment:
This call is already made on line 156. Could we reuse the CloudApplication?

Reply:
Done.